module/data/fields.mjs

```diff
@@ -0,0 +1,91 @@
+/**
+ * Data field that automatically selects the Advancement-specific configuration or value data models.
+ *
+ * @param {Advancement} advancementType  Advancement class to which this field belongs.
+ */
+export class AdvancementDataField extends foundry.data.fields.ObjectField {
+  constructor(advancementType, options={}) {
+    super(options);
+    this.advancementType = advancementType;
+  }
+
+  /** @inheritdoc */
+  static get _defaults() {
+    return foundry.utils.mergeObject(super._defaults, {required: true});
+  }
+
+  getModel() {
+    return this.advancementType.metadata?.dataModels?.[this.name];
+  }
+
+  getDefaults() {
+    return this.advancementType.metadata?.defaults?.[this.name] ?? {};
+  }
+
+  _cleanType(value, options) {
+    if ( !(typeof value === "object") ) value = {};
+
+    // Use a defined DataModel
+    const cls = this.getModel();
+    if ( cls ) return cls.cleanData(value, options);
+    if ( options.partial ) return value;
+
+    // Use the defined defaults
+    const defaults = this.getDefaults();
+    return foundry.utils.mergeObject(defaults, value);
```

Comment:
I missed this, but `inplace: true` is the default so this is still mutating the `defaults`. We need to explicitly pass `inplace: false`.